core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java

```diff
@@ -137,25 +137,24 @@ private void waitForMinimumWorkers()
 
     private void startExecution(QueryExecution queryExecution)
     {
-        queryExecutor.execute(() -> {
-            if (stateMachine.transitionToDispatching()) {
-                try {
-                    querySubmitter.accept(queryExecution);
-                    if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
-                        queryExecution.addFinalQueryInfoListener(queryMonitor::queryCompletedEvent);
-                    }
-                }
-                catch (Throwable t) {
-                    // this should never happen but be safe
-                    stateMachine.transitionToFailed(t);
-                    log.error(t, "query submitter threw exception");
-                    throw t;
-                }
-                finally {
-                    submitted.set(null);
+        if (stateMachine.transitionToDispatching()) {
+            try {
+                querySubmitter.accept(queryExecution);
+                if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
+                    queryExecution.addFinalQueryInfoListener(queryMonitor::queryCompletedEvent);
                 }
+                queryExecutor.execute(() -> queryExecution.start());
```

Comment:
I think this breaks the abstraction between `LocalDispatchQuery` and `SqlQueryManager`.  The SqlQueryManager has been responsible for coordinating the start of queries after and making sure that the interactions with the query tracker are handled property. 

I also don't understand why moving the start here helps with the root issue of reporting query state.

Reply:
The issue is that by calling `queryExecution.start()` synchronously, the `submitted` future won't be completed until _after_ it returns which means that it won't happen until after planning completes. That future controls when the client will be redirected to the `ExecutingStatementResource` from `QueuedStatementResource` which prevents the client from observing `PLANNING` as a state. The other reordering is incidental but necessary to ensure that the callbacks are all registered appropriately before the potential failure of the query execution could occur (likely from inside `start()`).

Reply:
@dain I think it doesn't necessarily break the abstraction. Because we can say that query submission is for preparing query execution(e.g. createQuery) and `queryExecution.start()` is for starting query execution. However, if you prefer the current level of abstraction, I can change SqlQueryManager instead to call `start()` using executor. But, with that, as James mentioned, `queryExecution.addFinalQueryInfoListener` will be added after the asynchronous `start()` call which would be more appropriate to be added before `start()`. Let me know what you think.